mux/route.go

```diff
@@ -6,7 +6,8 @@ import (
 
 // Route is an middleware aware route interface
 type Route interface {
-	Handler() interface{}
-	AppendMiddleware(m middleware.Middleware)
+	Handler(path string) interface{}
```

Comment:
I think the Route interface should stay the same, what we want to change would be the 
Middleware and Tree/Node so we can add to the tree Middleware same way how we add Nodes
And then when traversing the tree just collect middleware until you find the node, append/prepend middleware to route, compute handler, and return it

Reply:
In my humble opinion there is a missing logic to cross-check each middleware path against request path when node.Route().Handler().(fasthttp.RequestHandler)(ctx) is called based on which to decide for example include m1, skip m2 and so on. Right now the middleware is merged no matter if there is a match or not because it's been composed before the request. I agree the Route interface should stay the same but we need to associate a middleware with a path in route.USE call and that is the problem I am facing right now since this will require a map[string]MiddlewareFunc which would allow to do the checks in route.go Handler for example and call Compose with two params - path and middleware. The problem is the map is fixed size  and we need the dynamic size of the slice...

Reply:
Yes you are right, but having a map is not a solution here. You want to move `middleware` from `Route` struct into the node tree

```diff
type staticNode struct {
	name     string
	children Tree

	route Route
+	middleware middleware.Middleware

	maxParamsSize uint8
	skipSubPath   bool
}
```
and then as I suggested select all middleware that should be applied to that route while finding it.

How it works right now is a bit different to what I think you think. Right now we map a middleware path to a route declared path (`/a/{param}`), not a request path (`/a/test`).

The thing we want to remember here are:
- If middleware was applied to `/a` you want it to be applied to all children routes (like `/a/b`, `/a/c`, etc)
- middleware should be executed in order they were applied (right now this order is respected)

So if we move the middleware from route https://github.com/vardius/gorouter/blob/9233b0485a17c45a4f5e443830a2c6fc27b0a5c9/route.go#L8 to tree node  https://github.com/vardius/gorouter/blob/9233b0485a17c45a4f5e443830a2c6fc27b0a5c9/mux/node.go#L65 we will be able to compute the handler at the same time while traversing the node tree to find a match 
https://github.com/vardius/gorouter/blob/9233b0485a17c45a4f5e443830a2c6fc27b0a5c9/mux/node.go#L75
https://github.com/vardius/gorouter/blob/9233b0485a17c45a4f5e443830a2c6fc27b0a5c9/nethttp.go#L130

Reply:
Check this out, I'd like to have it done this way, not all tests pass right now, but it is a good draft. Later one we can cache computed handler. For example add method to router `.Compile()` where developer can call it once he knows all routes and middleware have been created. So he can precompute middleware per node. This way it doesn't have to be done on each request) and we will not loose much performance.

https://github.com/vardius/gorouter/pull/15/commits/a2896ff481c84fb0490bfb0e71e2604b93bb539a

Reply:
Yeah you are right now I understand the logic behind it better. Forked again from your latest changes and we can take it from there.